### PR TITLE
Ez copy data selections

### LINKED ISSUE #723: Copy data selections to another workspace
Allow users to copy selections of table data from one workspace to another. When a user makes a multiselection as per #404, provide a "Copy to another workspace" option. This should copy those entities to another workspace. If the entities have dependencies, the parents should be copied as well. This should have existing backend support, as it is available in FireCloud.

There, it is implemented as a pull. I am suggesting we change this to a push.

### PR DESCRIPTION
**Pull request by elizzhao**
Fixes #723 
Note: does not deal with select all across pages yet

Behavior:
Select and choose copy data
-if direct hard conflicts, asks user if they'd like to delete the existing (and potentially anything that references the existing such as sample sets, etc.) and override with the selected
-if soft conflicts, asks user if they'd like to link the selected entities to the existing 

Particularly looking for ways to improve:
-ExportDataModal: I use a lot of state and there may be better ways to write pieces of the code
-Warning/Alert messaging to user

### CODE REVIEW

**Comment by elizzhao** [comment]
Note: this implementation is good enough for now, we will revisit it later when our desired update semantics are better understood.

File: src/components/ExportDataModal.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+      ...warningStyle,
+      border: `1px solid ${colors.red[1]}`,
+      backgroundColor: colors.red[4],
+      color: colors.red[0]
+    }
+
+    const errors = validate(
+      { selectedWorkspaceId },
+      { selectedWorkspaceId: { presence: true } },
+      { prettify: v => (validate.prettify(v)) }
```
Comment (panentheos) [inline src/components/ExportDataModal.js]: nit: You only have to specify this option if you're overriding the default prettify function somehow.

File: src/components/ExportDataModal.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    super(props)
+
+    this.state = {
+      hardConflicts: [],
+      softConflicts: [],
+      additionalDeletions: [],
+      selectedWorkspaceId: undefined,
+      error: undefined,
+      copying: false,
+      selectedEntityType: props.selectedDataType
```
Comment (panentheos) [inline src/components/ExportDataModal.js]: This doesn't have to be in state since it doesn't change, so you could just use the prop instead.

File: src/components/ExportDataModal.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+import { reportError } from 'src/libs/error'
+import { requiredFormLabel, formLabel } from 'src/libs/forms'
+import * as Nav from 'src/libs/nav'
+import * as Style from 'src/libs/style'
+import * as Utils from 'src/libs/utils'
+import { Component } from 'src/libs/wrapped-components'
+import validate from 'validate.js'
+
+
+class infoTile extends Component {
```
Comment (panentheos) [inline src/components/ExportDataModal.js]: This is a good abstraction. A few nits:
Since it's stateless, you could use the 'functional' component notation which is slightly more compact.
PropTypes are probably overkill for something simple and file-local like this.
Component names should generally be title-cased, e.g. `InfoTile`

File: src/components/ExportDataModal.js [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+            additionalDeletions: _.filter(entity => entity.entityType !== selectedEntityType,
+              await error.json()), copying: false
+          }) //handles dangling references when deleting entities
+          return
+        default:
+          reportError('Error deleting data entries', error)
+          onDismiss()
+      }
+    }
+    this.copy()
```
Comment (panentheos) [inline src/components/ExportDataModal.js]: The logic is well de-duplicated, but right now `copy` and `doOverride` are co-recursive, and rely on the state change to terminate. Even if this works now, it's risky because there's no guarantee that the results of a `setState` will be visible during the current tick. What about combining everything into a single function with a flow like:
```js
async copy() {
  if (hardConflicts.length) {
    // delete entries
  }
  // copy entries
}
```
with a couple bail-out points for stopping early to report conflicts.

File: src/pages/workspaces/workspace/Data.js [hunk 5]
```diff
@@ -510,9 +513,18 @@ const EntitiesContent = ajaxCaller(class EntitiesContent extends Component {
           })
         ])
       ]),
+      !!selectedEntities.length && h(FloatingActionButton, {
```
Comment (breilly2) [inline src/pages/workspaces/workspace/Data.js]: In #930, I changed the type of `selectedEntities` from an array to an object. You should be able to change this to `_.isEmpty(selectedEntities)` now and it will work with either type.

File: src/pages/workspaces/workspace/Data.js [hunk 6]
```diff
@@ -525,6 +537,11 @@ const EntitiesContent = ajaxCaller(class EntitiesContent extends Component {
         namespace, name,
         selectedEntities, selectedDataType: entityKey, runningSubmissionsCount
       }),
+      copyingEntities && h(ExportDataModal, {
+        onDismiss: () => this.setState({ copyingEntities: false }),
+        workspace,
+        selectedEntities, selectedDataType: entityKey, runningSubmissionsCount
```
Comment (breilly2) [inline src/pages/workspaces/workspace/Data.js:543]: Also needs to change due to #930. Will need to be `selectedEntities: _.keys(selectedEntities)`. Whether now or later depends on which PR merges first. If this merges first, I'll take care of the change as part of #930.

File: src/components/FloatingActionButton.js [hunk 7]
```diff
@@ -43,7 +45,7 @@ export default class FloatingActionButton extends Component {
             transition: 'max-width 0.5s ease-out, padding 0.1s linear 0.2s'
           }
         }, label),
-        icon(iconShape, { size: 25, style: { stroke: 'white', strokeWidth: 2 } })
+        icon(iconShape, { size: 25, style: { stroke: 'white', strokeWidth: 0.25 } })
```
Comment (breilly2) [inline src/components/FloatingActionButton.js]: Was this intentional? I don't notice any difference on my screen.

Comment (elizzhao) [reply src/components/FloatingActionButton.js]: I initially had made it thinner when I was trying to use the three-dot card menu since the wide thickness looked weird with that icon. However, since we're doing the two buttons for now instead, I'll change it back since it looks fine both ways!

**Review by panentheos — APPROVED** [review]
Looks good.

File: src/components/ExportDataModal.js [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    const errors = validate(
+      { selectedWorkspaceId },
+      { selectedWorkspaceId: { presence: true } },
+    )
+
+    return h(Modal, {
+      onDismiss,
+      title: 'Copy to Workspace',
```
Comment (katevoss) [inline src/components/ExportDataModal.js:87]: Suggestion: "Copy Data to Workspace"?

File: src/components/ExportDataModal.js [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+      okButton: buttonPrimary({
+        tooltip: (hardConflicts.length !==0) ? 'Override existing entities, are you sure?' : Utils.summarizeErrors(errors),
```
Comment (katevoss) [inline src/components/ExportDataModal.js:89]: I would rephrase as "Are you sure you want to override existing data?"

File: src/components/ExportDataModal.js [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+      h(WorkspaceSelector, {
+        workspaces: _.filter(({ workspace: { workspaceId }, accessLevel }) => {
+          return workspace.workspaceId !== workspaceId && Utils.canWrite(accessLevel)
+        }, workspaces),
+        value: selectedWorkspaceId,
+        onChange: v => this.setState({ selectedWorkspaceId: v })
+      }),
+      (hardConflicts.length !== 0) && InfoTile({
+        infoStyle: errorStyle, iconName: 'error-standard',
+        content: 'The following entries already exist in the selected workspace. Please select CANCEL to go back or COPY to override the existing entities. '
```
Comment (katevoss) [inline src/components/ExportDataModal.js:109]: How about: "Some of the following data already exists in the selected workspace. Click CANCEL to go back or COPY to override the existing data."
Thoughts:
- I think we want to avoid using the word "entities" as it doesn't mean much to users. 
- If a user selects 4 pieces of data and 2 pieces are duplicates, they will see this message. Correct? In which case I'd rephrase to "Some of"
- IMO a user _selects_ data and _clicks_ COPY. It's a semantic note but I figured I'd put it in while I'm here :)

File: src/components/ExportDataModal.js [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+      }),
+      moreToDelete && InfoTile({
+        infoStyle: warningStyle, iconName: 'warning-standard',
+        content: 'In order to override the selected data entries, the following entries that reference the original data must ALSO be deleted.'
```
Comment (katevoss) [inline src/components/ExportDataModal.js:113]: To clarify the UX, if a user deletes a data entry that another data entry references, the ref will be deleted. Is that correct?
In which case I'd recommend simplifying the message: "To override the selected data entries, the following entries that reference the original data will also be deleted."

File: src/components/ExportDataModal.js [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+      }),
+      (softConflicts.length !== 0) && InfoTile({
+        infoStyle: warningStyle, iconName: 'warning-standard',
+        content: 'The following is linked to entries which already exist in the selected workspace. You may link the following to the existing entities by clicking COPY.'
```
Comment (katevoss) [inline src/components/ExportDataModal.js:117]: This feels like it's missing a word, like "The following data is linked...". Also again note use of entities, which I don't know what it means to users.

**Comment by katevoss** [comment]
Sorry to be adding comments late, but I noticed a few things:
- When I select entities to copy and click "Copy", the popup modal gives a dropdown list of workspaces to copy the data to. This dropdown includes the current workspace. Shouldn't we only include workspaces aside from the current workspace?
- If a user tries to copy duplicate data I get the message "The following entries already exist in the selected workspace. Please select CANCEL to go back or COPY to override the existing entities." While that message is in the modal, I can change the workspace destination. But that doesn't clear the message. When I click "COPY" I don't know whether I am copying data to the first workspace I selected, or the one I changed to.
- Added suggestions inline for wording

**Comment by elizzhao** [comment]
> * When I select entities to copy and click "Copy", the popup modal gives a dropdown list of workspaces to copy the data to. This dropdown includes the current workspace. Shouldn't we only include workspaces aside from the current workspace?

This seems to be the behavior for all workspace selection drop-downs that we have. It makes sense that we shouldn't let the user select the current workspace in this case, but not sure if we should make an entirely new workspace selector (could make sense for all the other "copy to another workspace" mechanics, but warrants discussion).

**Comment by elizzhao** [comment]
> * If a user tries to copy duplicate data I get the message "The following entries already exist in the selected workspace. Please select CANCEL to go back or COPY to override the existing entities." While that message is in the modal, I can change the workspace destination. But that doesn't clear the message. When I click "COPY" I don't know whether I am copying data to the first workspace I selected, or the one I changed to.

So the logic around here is pretty delicate (and something we definitely want to improve on in the future) and wouldn't handle selecting a new workspace while you've already started down one of the conflict paths. For now, the solution I'm implementing is just removing the workspace selector drop-down menu after the first "Copy" click. 

> * Added suggestions inline for wording

Added these in, thanks! They're a lot clearer now :)
